### ISSUE TITLE
Add supports method to formatters

### ISSUE THREAD

**Issue by sagikazarmark** [moneyphp/money#205]
We have the same case with the Bitcoin formatter: decorator.

Although it is not a good design to include a `supports` method just to make it compatible with some sort of AggregateMoneyFormatter, we should consider doing so. We might be able to add custom Formatters as well, based on different conditions.

Alternatively we could also have a SelectiveMoneyFormatter interface with it which could be used in an AggregateMoneyFormatter.

**Comment by frederikbosch** [comment]
@sagikazarmark Should this not belong in the client library? The client library is responsible for injecting the correct formatter into a class that is using a formatter. This resonsibility should not be passed to the formatter itself I think.

**Comment by sagikazarmark** [comment]
Well, I agree with you and I have some doubt in myself. However the current formatter tries to provide a fallback logic by accepting a fallback formatter and I was thinking if we could remove the decoration somehow.

**Comment by frederikbosch** [comment]
Ah I get it. You mean the delegatedFormatter inside the Bitcoin formatter? It might be a good idea indeed to remove that. Let me think on this.

**Comment by sagikazarmark** [comment]
Aye. Also, what about the parser? We can't really have a supports method there. Maybe we could have an AggregateParser based and determine "support" based on the return value (money object or not).

**Comment by frederikbosch** [comment]
@sagikazarmark What about something like this?

```php
<?php
final class CurrencySpecificMoneyFormatter implements MoneyFormatter {
  private $formatters = [];
  public function __construct(array $formatters) {
    foreach ($formatters as $currencyCode => $formatter) {
      $this->addFormatter($currencyCode, $formatter);
    }
  }

  private function addFormatter($currencyCode, MoneyFormatter $formatter) {
    $this->formatters[$currencyCode] = $formatter;
  }

  public function format(Money $money) {
    $currencyCode = $money->getCurrency()->getCode();
    if (isset($this->formatters[$currencyCode])) {
      return $this->formatters[$currencyCode]->format($money);
    }
    throw new FormatterException('Cannot format currency ' . $currencyCode);
  }
}
```

**Comment by frederikbosch** [comment]
Or with a default.

```php
<?php
final class AggregateMoneyFormatter implements MoneyFormatter {
  private $currencySpecificFormatters = [];
  private $defaultFormatter;
  public function __construct(MoneyFormatter $defaultFormatter, array $currencySpecificFormatters) {
    $this->defaultFormatter = $defaultFormatter;
    foreach ($currencySpecificFormatters as $currencyCode => $formatter) {
      $this->addFormatter($currencyCode, $formatter);
    }
  }

  private function addFormatter($currencyCode, MoneyFormatter $formatter) {
    $this->currencySpecificFormatters[$currencyCode] = $formatter;
  }

  public function format(Money $money) {
    $currencyCode = $money->getCurrency()->getCode();
    if (isset($this->currencySpecificFormatters[$currencyCode])) {
      return $this->currencySpecificFormatters[$currencyCode]->format($money);
    }
    return $this->defaultFormatter->format($money);
  }
}

**Comment by frederikbosch** [comment]
Then there is no need for `supports` method. Because the aggregate formatter is handling which formatter should be handled by which money object. In this case the currency code is de determinative factor. Clients can implement their own strategy with another determinative factor.

**Comment by sagikazarmark** [comment]
Looks good to me. Maybe we could add the default as part of the formatter array with a key "default" or "*". Another thing that we might think about is adding regex support for keys so that it is easier to register one formatter for many currencies.

**Comment by frederikbosch** [comment]
Alright, we could do the `*` method. Is regex the method to match currencies?

**Comment by sagikazarmark** [comment]
Yeah, regex could match currencies, but not sure it's the way we should do it. It would after all be the same as adding the formatter multiple times (listing the currencies) and adds a considerable amount of overhead, so let's drop regexes.

**Comment by frederikbosch** [comment]
Closed by #216
